### DISCUSSION TITLE
Query DuckDB using SQL?

### DISCUSSION THREAD

**Discussion by braddo99** [duckdb/duckdb#13862] [Q&A]
Hi all. I think I understand the primary purpose of DDB is as an embedded database, especially powerful because of its Python integration. I'm interested in DDB as a portable solution due to its install simplicity and ability to query text files so that I don't actually have to manage a database but could emulate some of its behaviors. My use case is to provide a place to house example data that one might query from a database. But, ... is it possible to query DDB itself using SQL? I mean, can I point a driver at it and use my application to query data? It's difficult to search for answers to this question because yes yes yes DDB does SQL and is great at it, but mostly DDB itself issuing the SQL, which is not my use case.

I've done a bunch of googling around and the answer seems to be no, which is puzzling to me even though I understand that's not its primary purpose. If one can send SQL using Python, it seems like it should be possible/(easy?) also to expose the SQL interface to an outside app. I'm not a developer myself but have access to some, so forgive my ignorance here.

If querying directly using SQL is not supported out of the box, is there a reference example of a Python app that can react to a driver, and just pass the SQL through to the embedded DDB?

Thanks for any illumination!

**Comment by soerenwolfers** [comment]
Is https://pypi.org/project/duckdb-server/ what you're looking for?

**Reply by braddo99** [reply]
Thanks for the reference, but it seems DDB server is optimized to respond to RESTful queries using get/post etc? I'm looking for a driver based system. I.e. I would like to install (or better using an existing common driver such as for MySQL etc) and, from my app (that doesn't contain DDB, it's an external application), connect to and query from DDB using SQL. This would be the standard way someone would connect to and query from say SQL Server, PostGRES, MySQL etc etc.  I love the idea of DDB as sort of a micro data lake (data pond?) so I can have a portable "database" with data in it, just zip it and go, nothing to install. But it would need to be connectable in the same way that classic databases are, not exclusively connected using APIs.

**Reply by soerenwolfers** [reply]
ah, then maybe https://duckdb.org/docs/api/odbc/overview.html is what you're after?

**Comment by robotslacker** [comment]
https://github.com/robotslacker/slackerdb   this can help you?
You can use JDBC to connect it.

I guess ODBC and python also work (because they are same PG protocol)

You can try it.  Let me know if you have any problem .  I'm the owner of it.

**Reply by braddo99** [reply]
thanks for the suggestion. At the moment I'm trying out MariaDB Portable and it seems to work in the way I wanted.

**Comment by fanyang01** [comment]
[MyDuck Server](https://github.com/apecloud/myduckserver) is designed to run DuckDB in server mode, supporting both MySQL and PostgreSQL protocols and drivers. It also offers data replication from MySQL and PostgreSQL. Let me know if you find it useful or if there's anything we could improve.

**Reply by robotslacker** [reply]
I have noticed "MyDuck Server" project, it's very very amazing.
Slackerdb is just a tool of mine, I worked it out when I was free.
I may continue improvements in the future, mainly focus on workflow and CDC.

Although I don't have much time and energy to modify it at this moment.
When I have time, I will study your project carefully, there are many things worth learning.

**Reply by fanyang01** [reply]
Thanks for your kind words. Feel free to try it out when you have time, and best of luck with SlackerDB — looks like a great tool for Java world! 😊
